[CALCITE-2417] Fix ClassCastException in RelToSqlConverter with structs

BenoitHanotte opened the pull request:
Trying to convert the pysical plan of a select * query on a table with a nested struct throws ClassCastException as RexInputRef cannot be cast to RexCorrelVariable.
This is due to the cast to RexCorrelVariable not being done on the referenced expression of RexInputRef  accessible through RexFieldAccess.getReferenceExpr()

I am not sure how you would advise testing such a case? In order to isolate the issue I had to recreate a  custom schema with a nested struct, however this is adds a lot of boiler plate.

michaelmior commented:
Overall LGTM. Can you rename `RelToSqlNestedStruct` to `RelToSqlNestedStructTest`? It would also probably be a good idea to validate the result of `RelToSqlConverter`.

zabetak commented:
Hi @BenoitHanotte, for creating appropriate tests I believe you can add a few test methods in `RelToSqlConverterTest`. A schema with some minimal nesting that may be sufficient for your case exists in `CalciteAssert.SchemaSpec.HR`. 

If you need a Schema with more nesting in order to test your feature wait for [PR762](https://github.com/apache/calcite/pull/762).

BenoitHanotte commented:
Thanks @zabetak , indeed the `employees` has a nested struct that allows testing this fix with a much reduced boiler plate. 
I updated my PR to add a test case to `RelToSqlConverterTest` and remove the `RelToSqlConverterNestedStructsTest` file, it is much cleaner this way :)

BenoitHanotte commented:
Unfortunately I discovered that this fix is not giving correct names for the nested fields:
```
SELECT * FROM depts
```
gives the following query:
```
SELECT deptno, name, employees, location, location AS location FROM hr.depts
```
I'll have alook at it this afternoon but from my limited understanding of the code we'll have to make sure that `RexFieldAccess` is actually going through the correlation path to keep track of the reference

BenoitHanotte commented:
@zabetak @michaelmior  I updated my PR, it seems that in addition to the cast issue we were not correctly mapping the field index to the `RelDataTypeField` of the row when the row type is nested.

For instance, for a `RecordType(INT a, RecordType(INT b)`, we need to build an ordered list such as
```
0: RelDataTypeField of a
1: RelDataTypeField of a.b
```
This way we can get the `RelDataTypeField` for given field index in an `AliasContext`

I added a few test cases to ensure that names were correctly retrieved for nested fields.

michaelmior commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -2658,6 +2692,7 @@ String exec() {
         for (Function<RelNode, RelNode> transform : transforms) {
           rel = transform.apply(rel);
         }
+        System.out.println(RelOptUtil.toString(rel));
```
Minor, but looks like this was for debugging and should be removed.

BenoitHanotte replied:
Sorry, I forgot to remove it

zabetak commented on core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:
```diff
@@ -441,11 +444,17 @@ public SqlNode toSql(RexProgram program, RexNode rex) {
+          final RexCorrelVariable variable = (RexCorrelVariable) access.getReferenceExpr();
+          final Context aliasContext = correlTableMap.get(variable.id);
+          return aliasContext.field(access.getField().getIndex());
+        default:
```
The code assumes that FIELD_ACCESS wraps always a CORREL_VARIABLE or INPUT_REF. I have the impression that this code will lead again to a ClassCastException just a bit later (e.g., for a doubly nested struct where there is a RexFieldAccess(RexFieldAccess(RexInputRef)).

BenoitHanotte replied:
You're right, I thought that it could only go down one level but after adding tests for multiple nested levels I see that it can happen, I am updating the PR to support this and add tests for this case

BenoitHanotte commented:
I updated the PR to do the following:
1. Support more than one level of RexFieldAccess (use a stack called `accesses` to store the `RexFieldAccess` levels, get the referenced `SqlNode`and then poll from the stack to add the name components)
2.  Add tests on a schema containing more than one level of nesting and more than one nested column

I haven't been able to find a more elegant solution to the correlation variable issue as I don't seem to be able to get the field index without having access to the last level of `RexFieldAccess`, hence the inner switch case to handle this particular case.

I tested this patch on our production schema and it seems to have solved the issue.

BenoitHanotte commented:
Hello @michaelmior @zabetak would you have time for a review of this ticket so that we can see whether this is the right approach or not? Thanks!

michaelmior commented:
LGTM, merged! Sorry for the delay. I was waiting for the latest release to go through and this slipped off my radar. Thanks for the contribution :)
